### ISSUE TITLE
[feature request] IDE without a server

### ISSUE THREAD

**Issue by trusktr** [eclipse-theia/theia#8368]
It'd be great to use [wasm-git](https://github.com/petersalomonsen/wasm-git) so that Theia can run on static websites without the need for a server. It can use the filesystem provided by wasm-git, without many modifications to Theia (if any), and without having to wait for VS Code's browser plugins.

This would be excellent because:

- Theia would run offline
- static demos (f.e. on codepen or anywhere similar)
- the git file system could be synced to any git repo (introduce saving of projects from static websites to actual servers when ready, incrementally upgrading a Theia-based project)

**Comment by trusktr** [comment]
See a usage example: https://github.com/petersalomonsen/wasm-git/blob/master/examples/example_webworker.js

**Comment by vince-fugnitto** [comment]
@trusktr thank you for the issue, unfortunately we do not believe supporting a serverless setup in the framework is feasible with the current architecture and will require too much effort, maintenance, testing overall.

In addition, there are no strong use-case for the community to additionally support serverless applications, and the linked `wasm-git` does not seem mature enough to be included as part of the solution.

However, if this is something you're interested in personally then you're always free to implement it, maintain it and use it for your own application(s).

Here are some additional issues with reasons behind supporting serverless setups:
- https://github.com/eclipse-theia/theia/issues/6560
- https://github.com/eclipse-theia/theia/issues/6732

--- 

> Theia would run offline

We already support a completely local setup (using `electron`) similarly to vscode.

**Comment by akosyakov** [comment]
@vince-fugnitto agree with you, but let's keep it, VS Code is moving in the direction running only in the browser too with rewrite of built-in and popular extensions. At some point of time it would make sense for us to follow.

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto agree with you, but let's keep it, VS Code is moving in the direction running only in the browser too with rewrite of built-in and popular extensions. At some point of time it would make sense for us to follow.

@akosyakov I might have misunderstood, but what would we need to do to follow? Are you saying we'd only be targeting `browser` in the future and `electron` will be dropped?

**Comment by akosyakov** [comment]
No, I mean there will be a mode when Theia only works in the browser without the server (offline mode) and then it can switch to online mode. Nothing will be dropped.

**Comment by JonasHelming** [comment]
We have initial support for this mode now. I am closing this. See linked PR and follow-ups.
